packages/create-amplify/src/package-manager-controller/package_manager_controller_factory.ts

```diff
@@ -0,0 +1,203 @@
+      return 'npm';
+    }
+
+    const packageManagerAndVersion = this.userAgent.split(' ')[0];
+    const packageManagerName = packageManagerAndVersion.split('/')[0];
+
+    if (packageManagerName === 'yarn') {
+      const yarnMajorVersion = packageManagerAndVersion
+        .split('/')[1]
+        .split('.')[0];
+      const yarnName = `${packageManagerName}-${
+        yarnMajorVersion === '1' ? 'classic' : 'modern'
+      }`;
+      return yarnName;
+    }
+    return packageManagerName;
+  }
+
+  /**
+   * getPackageManagerController
+   */
+  getPackageManagerController() {
+    const packageManagerName = this.getPackageManagerName();
+    switch (packageManagerName) {
+      case 'npm':
+        return new NpmPackageManagerController(
+          this.projectRoot,
+          new PackageManagerControllerFactory(this.projectRoot, this.userAgent)
+        );
+      case 'pnpm':
+        return new PnpmPackageManagerController(
+          this.projectRoot,
+          new PackageManagerControllerFactory(this.projectRoot, this.userAgent)
+        );
+      case 'yarn-classic':
+        return new YarnClassicPackageManagerController(
+          this.projectRoot,
+          new PackageManagerControllerFactory(this.projectRoot, this.userAgent)
+        );
+      case 'yarn-modern':
+        return new YarnModernPackageManagerController(
+          this.projectRoot,
+          new PackageManagerControllerFactory(this.projectRoot, this.userAgent)
+        );
+      default:
+        throw new Error(
+          `Package Manager ${packageManagerName} is not supported.`
+        );
+    }
+  }
+
+  /**
+   * initializeProject
+   */
+  async initializeProject(packageManagerProps: {
+    executable: string;
+    initDefault: string[];
+  }) {
+    if (this.packageJsonExists(this.projectRoot)) {
+      // if package.json already exists, no need to do anything
+      return;
+    }
+
+    logger.debug(
+      `No package.json file found in the current directory. Running \`${packageManagerProps.executable} init\`...`
+    );
+
+    try {
+      await this.executeWithDebugLogger(
+        this.projectRoot,
+        packageManagerProps.executable,
+        packageManagerProps.initDefault,
+        this.execa
+      );
+    } catch {
+      throw new Error(
+        `\`${packageManagerProps.executable} init\` did not exit successfully. Initialize a valid JavaScript package before continuing.`
+      );
+    }
+
+    if (!this.packageJsonExists(this.projectRoot)) {
+      // this should only happen if the customer exits out of npm init before finishing
+      throw new Error(
+        `package.json does not exist after running \`${packageManagerProps.executable} init\`. Initialize a valid JavaScript package before continuing.'`
+      );
+    }
+  }
+
+  /**
+   * Copies the template directory to an amplify folder within the projectRoot
+   */
+  generateInitialProjectFiles = async (packageManagerProps: {
```

Comment:
this also shouldn't be on a factory.

I think the previous version that had a ProjectFileInitializer was a better factoring of this logic. The ProjectFileInitializer can initialize common files and be injected with a PackageManagerController to initialize package manager specific files.

Reply:
Done